File: src/main/java/com/nanaios/polygonal_tech/capability/energy/EmptyLongEnergyStorage.java
package com.nanaios.polygonal_tech.capability.energy;

import com.nanaios.polygonal_tech.capability.interfaces.ILongEnergyStorage;

/// ILongEnergyStorageの空実装。常に0を返し、エネルギーの挿入や抽出を受け付けません。
/// ifPresentのデフォルト値として使用できます。
/// 例えば、エネルギー貯蔵が存在しない場合にこのクラスのインスタンスを返すことで、呼び出し側はnullチェックをせずに安全にエネルギー貯蔵を操作できます。
/// このクラスはシングルトンパターンで実装されており、INSTANCEフィールドを通じて唯一のインスタンスにアクセスできます。
public class EmptyLongEnergyStorage implements ILongEnergyStorage {
    public static final EmptyLongEnergyStorage INSTANCE = new EmptyLongEnergyStorage();

    @Override
    public long getLongEnergyStored() {
        return 0;
    }

    @Override
    public long getLongMaxEnergyStored() {
        return 0;
    }

    @Override
    public long receiveLongEnergy(long maxReceive, boolean simulate) {
        return 0;
    }

    @Override
    public long extractLongEnergy(long maxExtract, boolean simulate) {
        return 0;
    }

    @Override
    public boolean canExtract() {
        return false;
    }

    @Override
    public boolean canReceive() {
        return false;
    }
}

File: src/main/kotlin/com/nanaios/polygonal_tech/core/capability/energy/EmptyLongEnergyStorage.kt
package com.nanaios.polygonal_tech.core.capability.energy

import com.nanaios.polygonal_tech.core.network.sync.storage.EmptySyncValueStorage
import com.nanaios.polygonal_tech.core.network.sync.storage.ISyncValueStorage
import com.nanaios.polygonal_tech.core.network.sync.type.ISyncType
import com.nanaios.polygonal_tech.core.network.sync.type.SyncType
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf

object EmptyLongEnergyStorage : ILongEnergyStorage {
    override val longEnergyStored = 0L
    override val maxLongEnergyStored = 0L
    override var storage: ISyncValueStorage = EmptySyncValueStorage
    override var type: ISyncType = SyncType.NONE
    override val isDirty = false

    override fun receiveLongEnergy(maxReceive: Long, simulate: Boolean) = 0L
    override fun extractLongEnergy(maxExtract: Long, simulate: Boolean) = 0L
    override fun canExtract() = false
    override fun canReceive() = false
    override fun writeBuffer(buffer: FriendlyByteBuf) = Unit
    override fun readBuffer(buffer: FriendlyByteBuf) = Unit
    override fun onSync() = Unit
    override fun serializeNBT() = CompoundTag()
    override fun deserializeNBT(nbt: CompoundTag) = Unit
}